frontend/manage_user.tsx

```diff
@@ -0,0 +1,220 @@
+/**
+ * @file This file contains the ManageUser function. This function is used for
+ * add users and set user permissions.
+ */
+import React from 'react'
+import { MenuItem, InputLabel, Select, TextField, Checkbox, FormControlLabel, Button } from '@material-ui/core';
+import { makeStyles } from '@material-ui/core/styles';
+import { get, post, update } from './communication';
+import { useDataContext } from './data_context'
+import { Genebank, Herd } from '~data_context_global';
+
+// Define styles for tab menu
+const useStyles = makeStyles({
+  form: {
+    borderLeft: "1px solid rgba(0,0,0,0.1)",
+    paddingLeft: "0.5cm",
+    display: "flex",
+    flexDirection: "column",
+  },
+  simpleField: {
+    width: "400px",
+  },
+});
+
+/** description of a user to manage */
+export interface ManagedUser {
+  id: number
+  email: string
+  validated: boolean
+  privileges: any
+  password?: string
+}
+
+/**
+ * Provides a form for changing user metadata and user roles.
+ */
+export function ManageUser({id}: {id: number | string | undefined}) {
+  const {genebanks, loadData} = useDataContext()
+  const [user, setUser] = React.useState(undefined as ManagedUser | undefined)
+  const [isNew, setNew] = React.useState(false)
+  const [level, setLevel] = React.useState("owner")
+  const [genebank, setGenebank] = React.useState(0)
+  const [herds, setHerds] = React.useState([] as Herd[])
+  const [herd, setHerd] = React.useState(0)
+  const classes = useStyles();
+
+  const loadUser = (id: number) => {
+    setNew(false)
+    get(`/api/manage/user/${id}`).then(
+      data => data && setUser(data),
+      error => console.error(error)
+    );
+  }
+
+  React.useEffect(() => {
+    if (typeof id == 'number') {
+      loadUser(id)
+    } else if (id == 'new') {
+      setNew(true)
+      setUser({id: -1, email: '', validated: false, privileges: []})
+    }
+  }, [id])
+
+  React.useEffect(() => {
+    if (genebanks.length > 0) {
+      selectGenebank(genebanks[0].id)
+    }
+  }, [genebanks])
+
+  const selectGenebank = (gId: number) => {
+    setGenebank(gId);
+    let genebankData = genebanks.filter(g => g.id == gId);
+    if (genebankData.length > 0) {
+      setHerds(genebankData[0].herds)
+      if (herds.length > 0 && herds.filter(h => herd == h.id).length == 0) {
+        setHerd(herds[0].id)
```

Comment:
Why do you first filter and then anyway set to the first elemnet?

Reply:
it might be a dumb way to write this check, but it's essentially "if there are herds, and the currently selected herd isn't in the current list of herds, then select the first herd as default" could be `herds.length > 0 && !herds.find(h => h.id == herd)` instead, and that might be nicer...

Reply:
Haha okej now I see it. I overlooked the `== 0` part but now that you pointed it out it's obvious of course. Yes `herds.length && !herd.find(...)` would be clearer. (Too bad `!` isn't `not` in JS though!)